query-parser/src/parse.rs

```diff
@@ -440,10 +511,23 @@ def_parser!(Find, aggregate_element, Element, {
     Query::aggregate().map(Element::Aggregate)
 });
 
+def_parser!(Find, pull_element, Element, {
+    seq().of_exactly(Query::pull().with(Query::variable().and(Query::pull_attributes())))
+         .map(|(var, attrs)| Element::Pull(Pull { var: var, patterns: attrs }))
+});
+
 def_parser!(Find, elem, Element, {
-    choice([try(Find::variable_element()),
-            try(Find::corresponding_element()),
-            try(Find::aggregate_element())])
+    choice([
+        try(Find::variable_element()),
```

Comment:
Think a little bit about which of these really need `try` -- maybe you can improve on the error messages we get.  For `corresponding_element`, you really want to commit to this branch as soon as you see `(the ...)`, and likewise for `(pull ...)`.  That's not expressed here, unfortunately.

Reply:
I rewrote this to use `look_ahead`.